File: mineSweeper/cell/cell.go
package cell

// Cell that populates the board
type Cell struct {
	Revealed      bool
	X             int
	Y             int
	Bee           bool
	Display       string
	NeighborCount int
}

func (c *Cell) floodFill(board [][]*Cell) {
	for xOff := -1; xOff <= 1; xOff++ {
		i := c.X + xOff
		if i < 0 || i >= len(board) {
			continue
		}

		for yOff := -1; yOff <= 1; yOff++ {
			j := c.Y + yOff
			if j < 0 || j >= len(board) {
				continue
			}

			neighbor := board[i][j]

			if !neighbor.Revealed {
				neighbor.Show(board)
			}
		}
	}
}

// Show sets the cell to Revealed
func (c *Cell) Show(board [][]*Cell) {
	c.Revealed = true

	if c.NeighborCount == 0 {
		c.floodFill(board)
	}
}

// CountBeeNeighbors counts the number of bees neighboring the cell
func (c *Cell) CountBeeNeighbors(board [][]*Cell) {

	if c.Bee {
		c.NeighborCount = -1
		return
	}

	total := 0

	for xOff := -1; xOff <= 1; xOff++ {
		i := c.X + xOff
		if i < 0 || i >= len(board) {
			continue
		}

		for yOff := -1; yOff <= 1; yOff++ {
			j := c.Y + yOff
			if j < 0 || j >= len(board) {
				continue
			}

			neighbor := board[i][j]

			if neighbor.Bee {
				total++
			}
		}
	}
	c.NeighborCount = total
}

File: ruby_programs/mineSweeper/cell.rb
class Cell
  
  attr_accessor :x, :y, :revealed, :display, :bee_neighbor_count, :bee
  
  def initialize(x, y)
    @x = x
    @y = y
    @revealed = false
    @display = "#"
    @bee_neighbor_count = 0
    @bee = false
  end

  def count(grid)

    if self.bee
      self.bee_neighbor_count = -1
    end

    bee_neighbor_count = 0

    (-1..1).each do |xOff|
      i = self.x + xOff
      next if i < 0 || i >= grid.length
      (-1..1).each do |yOff|
        j = self.y + yOff
        next if j < 0 || j >= grid.length

        neighbor = grid[i][j]

        if neighbor.bee
          bee_neighbor_count+=1
        end

      end
    end
    self.bee_neighbor_count = bee_neighbor_count
  end

  def reveal(grid)
    self.revealed = true
    if self.bee_neighbor_count == 0
      reveal_all_neighbors(grid)
    end
  end

  def reveal_all_neighbors(grid)
    (-1..1).each do |xOff|
      i = self.x + xOff
      next if i < 0 || i >= grid.length
      (-1..1).each do |yOff|
        j = self.y + yOff
        next if j < 0 || j >= grid.length

        neighbor = grid[i][j]

        if !neighbor.revealed
          neighbor.reveal(grid)
        end

      end
    end
  end


end
